In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

Berikut ini adaalah beberapa tools yang akan digunakan untuk melakukan regresi linear dan matriks pada dataset yang kita inginkan

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving dataset_satelit.csv to dataset_satelit (1).csv


Code ini untuk melakukan penguploadan file atau dataset yang akan dieksekusi

In [ ]:
df = pd.read_csv('dataset_satelit.csv')

No   Longitude  Lattitude     N     P      K    Ca    Mg      Fe      Mn  \
0   1  103.036658  -0.604417  2.64  0.15  0.415  0.51  0.31  119.96  463.23   
1   2  103.037201  -0.604689  2.75  0.17  0.568  0.76  0.58  102.63  493.81   
2   3  103.036359  -0.603012  1.77  0.12  0.339  0.49   0.6  107.37  460.93   
3   4  103.036950  -0.603219  2.30  0.15  0.460  0.74  0.67   96.02  338.17   
4   5  103.036802  -0.601969  2.05  0.14  0.308  0.64  0.72   87.01  384.33   

   ...      b1  Sigma_VV  Sigma_VH      plia       lia      iafe  gamma0_vv  \
0  ...  0.0433   0.18183   0.04461  35.74446  35.79744  35.41161    0.22331   
1  ...  0.0465   0.22079   0.04640  35.12096  35.14591  35.41510    0.27116   
2  ...  0.0417   0.18926   0.03992  35.07724  35.07730  35.41135    0.23242   
3  ...  0.0367   0.14769   0.03622  36.08078  36.08469  35.41583    0.18138   
4  ...  0.0361   0.18205   0.03797  32.68855  32.69293  35.41592    0.22359   

   gamma0_vh  beta0_vv  beta0_vh  
0    0.05479   0.31325   0.07686  
1    0.05699   0.38033   0.07993  
2    0.04902   0.32604   0.06876  
3    0.04448   0.25440   0.06238  
4    0.04664   0.31359   0.06541  

[5 rows x 34 columns]

Kode ini untuk memerintahkan pandas untuk membaca dataset yang sudah kita upload dan bisa kita lihat kolom' yang terdapat didalam dataset berikut

#inetrprentasi kolom


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 594 entries, 0 to 593
Data columns (total 34 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   No         594 non-null    int64  
 1   Longitude  594 non-null    float64
 2   Lattitude  594 non-null    float64
 3   N          594 non-null    float64
 4   P          594 non-null    float64
 5   K          593 non-null    float64
 6   Ca         594 non-null    float64
 7   Mg         594 non-null    object 
 8   Fe         594 non-null    float64
 9   Mn         594 non-null    float64
 10  Cu         594 non-null    float64
 11  Zn         594 non-null    float64
 12  B          594 non-null    float64
 13  b12        594 non-null    float64
 14  b11        594 non-null    float64
 15  b9         594 non-null    float64
 16  b8a        594 non-null    float64
 17  b8         594 non-null    float64
 18  b7         594 non-null    float64
 19  b6         594 non-null    float64
 20  b5        

#informasi datase
Bisa kita lihat data ini terdiri dari 594 row dan 33 kolom


In [ ]:
df.describe()

No   Longitude   Lattitude           N           P           K  \
count  594.000000  594.000000  594.000000  594.000000  594.000000  593.000000   
mean   297.500000  106.878644   -1.024933    2.259091    0.141380    0.582175   
std    171.617307    4.949840    0.965349    0.395499    0.019782    0.222567   
min      1.000000  102.760857   -2.333750    1.140000    0.090000    0.122000   
25%    149.250000  102.927811   -2.233338    1.982500    0.130000    0.429000   
50%    297.500000  103.581969   -0.602276    2.280000    0.140000    0.549000   
75%    445.750000  113.403797   -0.257349    2.570000    0.150000    0.710000   
max    594.000000  113.434700    0.069251    3.230000    0.220000    1.489000   

               Ca          Fe           Mn          Cu  ...          b1  \
count  594.000000  594.000000   594.000000  594.000000  ...  594.000000   
mean     0.595094   74.613771   308.034697    2.391195  ...    0.177291   
std      0.366118   55.579655   241.731643    1.580296  ...    0.155615   
min      0.050000   21.080000     3.160000    0.090000  ...    0.014100   
25%      0.320000   40.705000   124.015000    1.172500  ...    0.046925   
50%      0.540000   65.650000   239.445000    2.225000  ...    0.072700   
75%      0.790000   87.372500   434.990000    3.357500  ...    0.318900   
max      2.820000  559.100000  2009.320000    8.170000  ...    0.751400   

         Sigma_VV    Sigma_VH        plia         lia        iafe   gamma0_vv  \
count  594.000000  594.000000  594.000000  594.000000  594.000000  594.000000   
mean     0.234474    0.102789   28.640422   28.664891   28.609569    0.202587   
std      0.070516    0.112310   15.325347   15.380384   15.329170    0.104357   
min      0.115170    0.021460    0.127000    0.098600    0.026000    0.008700   
25%      0.183210    0.039535   31.959745   31.968948   33.685353    0.183085   
50%      0.213385    0.046550   35.067930   35.110415   34.611565    0.233590   
75%      0.262242    0.059190   38.319135   38.441065   39.002760    0.271790   
max      0.512210    0.373000   47.592900   48.014640   39.209330    0.658960   

        gamma0_vh    beta0_vv    beta0_vh  
count  594.000000  594.000000  594.000000  
mean     0.051524    0.269642    0.062320  
std      0.012959    0.143728    0.024218  
min      0.016900    0.009300    0.016300  
25%      0.040250    0.244935    0.052772  
50%      0.050415    0.310380    0.068380  
75%      0.060410    0.364505    0.079020  
max      0.122300    0.814170    0.150620  

[8 rows x 33 columns]

Kita lihat pada kolom Longitude dan lattitude, data ini mencakup wilayah geografis di sekitar Indonesia bagian barat (kemungkinan Sumatera dan Jawa), mengingat koordinatnya berada di sekitar garis khatulistiwa


In [ ]:
df.columns

Index(['No', 'Longitude', 'Lattitude', 'N', 'P', 'K', 'Ca', 'Mg', 'Fe', 'Mn',
       'Cu', 'Zn', 'B', 'b12', 'b11', 'b9', 'b8a', 'b8', 'b7', 'b6', 'b5',
       'b4', 'b3', 'b2', 'b1', 'Sigma_VV', 'Sigma_VH', 'plia', 'lia', 'iafe',
       'gamma0_vv', 'gamma0_vh', 'beta0_vv', 'beta0_vh'],
      dtype='object')

Memintaa program untuk menampilkan kolom' yang ada pada dataset yang sudahkita upload.

In [ ]:
#ubah kolom Mg ke float
df['Mg'] = pd.to_numeric(df['Mg'], errors='coerce')
# hilangkan baris yang ada missing value
df = df.dropna()

pada kode ini kita meminta kepada pandas untuk mengubah kolom Mg yang berawal type datanyaa yaitu object menjadi numerik apabila nanti ada nilai yang tidak sesuai maka akan berubah menjadii Nan dan semua missing value itu di hapus

In [ ]:
# X = df[['b2', 'b3', 'b4', 'b8', 'b11', 'Sigma_VV', 'Sigma_VH']]
# y = df['N']

In [ ]:
X = df[['P', 'K', 'Ca', 'Mg', 'Fe', 'Mn', 'Cu', 'Zn', 'B', 'b12', 'b11', 'b9', 'b8a', 'b8', 'b7', 'b6', 'b5',
        'b4', 'b3', 'b2', 'b1', 'Sigma_VV', 'Sigma_VH', 'plia', 'lia', 'iafe',
         'gamma0_vv', 'gamma0_vh', 'beta0_vv', 'beta0_vh']]
y = df['N']

Pada bagian ini kita menentukan variabel X dan Y, variabel X adalah variabel yang menentukan Fitur yang dimana akan memiliki korelasi terhadap variabel Y sebagaai variabel yang terikat

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Disini kita membagi data training dan data testing, data training 80% dan data testing 20%. X_train: Fitur untuk training (30 variabel prediktor),X_test: Fitur untuk testing, y_train: Target (Nitrogen) untuk training, y_test: Target untuk validasi model


In [ ]:
# training data
model = LinearRegression()

model.fit(X_train, y_train)

LinearRegression()

kode ini untuk menentukan model apa yang akan di gunakan, dan disini saya menggunakan Regresi Liniear, model fit ini digunakan untuk melatih data tersebut, Algoritma akan mencari koefisien optimal untuk setiap fitur. Tujuan: Meminimalkan **Mean Squared Error (MSE)** antara prediksi dan nilai aktual

In [ ]:
# testing
y_pred = model.predict(X_test)

r2 = r2_score(y_test, y_pred) #evaluasi semakin kecil semakin baik
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

print("RMSE:", rmse)
print("R2 Score :", r2)

RMSE: 0.19392193397227142
R2 Score : 0.7549955124484078


Disini kita mengetahui bahwasanyaa Rata' kesalahan prediksi sekitar **±0.194 unit nitrogen**
Mengingat rentang nitrogen dalam data:
  - Min: 1.14
  - Max: 3.23
  - Range: 2.09
  - Mean: 2.26
- **RMSE relatif = 0.194/2.09 ≈ 9.3%** dari rentang nilai
- **Kesalahan relatif terhadap mean = 0.194/2.26 ≈ 8.6%**

**Kesimpulan RMSE:**
- **Baik** - Model memiliki akurasi prediksi yang cukup tinggi
- Kesalahan ~8-9% dari nilai rata-rata adalah acceptable untuk data environmental/geospatial
- Untuk aplikasi praktis (rekomendasi pemupukan), akurasi ini sudah memadai

**Interpretasi:**
- Model dapat menjelaskan **75.5%** dari variabilitas kadar nitrogen
- **24.5%** sisanya dijelaskan oleh faktor lain yang tidak termasuk dalam model

**Penilaian R²:**
- 0.7-0.8: **Sangat Baik** untuk data environmental
- 0.5-0.7: Baik
- <0.5: Perlu improvement

**Kesimpulan R²:**
- **Sangat Baik** - Model memiliki kemampuan prediksi yang kuat
- Untuk dataset dengan 30+ fitur heterogen (lab + remote sensing), R² = 0.755 adalah hasil yang excellent
- Menunjukkan kombinasi data tanah dan satelit sangat efektif untuk prediksi nitrogen


In [ ]:
coeff = pd.DataFrame({
    'Fitur' : X.columns,
    'Koefisien' : model.coef_
})

print(coeff)

        Fitur  Koefisien
0           P   8.570837
1           K   0.056014
2          Ca  -0.058380
3          Mg  -0.161095
4          Fe   0.000284
5          Mn   0.000177
6          Cu   0.023202
7          Zn   0.002968
8           B  -0.002092
9         b12   0.327312
10        b11  -0.619437
11         b9  -0.047586
12        b8a   0.064453
13         b8  -0.072659
14         b7   0.981186
15         b6  -0.864947
16         b5  -0.329397
17         b4   1.420614
18         b3  -0.449043
19         b2  -0.357454
20         b1  -0.033418
21   Sigma_VV   0.156595
22   Sigma_VH  -1.466200
23       plia   0.021315
24        lia  -0.021906
25       iafe  -0.078122
26  gamma0_vv   0.457327
27  gamma0_vh   2.946416
28   beta0_vv  -0.569281
29   beta0_vh  -0.677599


Fitur dengan Pengaruh Positif Terkuat (Meningkatkan Prediksi):b4 ($\approx 1.4286$) memiliki dampak positif paling besar. Kenaikan pada b4 akan sangat meningkatkan nilai variabel yang diprediksi oleh model.P ($\approx 0.5788$) dan b7 ($\approx 0.9811$) juga memberikan kontribusi positif yang signifikan.Fitur dengan Pengaruh Negatif Terkuat (Menurunkan Prediksi):b6 ($\approx -0.8649$) memiliki dampak negatif paling besar. Kenaikan pada b6 akan sangat menurunkan nilai variabel yang diprediksi.b2 ($\approx -0.3574$) dan Mg ($\approx -0.1616$) juga memberikan kontribusi negatif.Fitur dengan Pengaruh Paling Kecil (Tidak Signifikan):Fe ($\approx 0.0001$) dan Cu ($\approx -0.0008$) memiliki koefisien yang sangat dekat dengan nol, menunjukkan bahwa variabel-variabel ini hampir tidak memiliki pengaruh pada hasil prediksi model.

In [ ]:
import statsmodels.api as sm

X_sm= sm.add_constant(X)
model_ols = sm.OLS(y, X_sm).fit()
print(model_ols.summary())

                            OLS Regression Results                            
Dep. Variable:                      N   R-squared:                       0.779
Model:                            OLS   Adj. R-squared:                  0.768
Method:                 Least Squares   F-statistic:                     66.10
Date:                Mon, 03 Nov 2025   Prob (F-statistic):          6.69e-163
Time:                        08:18:40   Log-Likelihood:                 156.68
No. Observations:                 592   AIC:                            -251.4
Df Residuals:                     561   BIC:                            -115.5
Df Model:                          30                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.8681      0.864      4.479      0.0

1. Performa Model
Model linear regression menunjukkan performa yang baik dengan R² = 77.9% (statsmodels) dan 75.5% (sklearn). Ini berarti model dapat menjelaskan sekitar 76-78% variabilitas kadar nitrogen dalam tanah. RMSE sebesar 0.194 menunjukkan rata-rata kesalahan prediksi sekitar 8-9% dari nilai rata-rata nitrogen, yang dapat diterima untuk aplikasi pertanian dan environmental monitoring.
2. Fitur Paling Penting
Dari 30 fitur yang digunakan, hanya 8 fitur yang signifikan secara statistik:
Sangat Signifikan (p < 0.01):

Fosfor (P): Prediktor terkuat dengan koefisien +8.576
Magnesium (Mg): Koefisien -0.217
Mangan (Mn): Koefisien +0.0002
Tembaga (Cu): Koefisien +0.023
Band 8/NIR (b8): Koefisien -0.081
Iron Absorption Index (iafe): Koefisien -0.080

Signifikan (p < 0.05):

Band 7/SWIR (b7): Koefisien +1.306
Band 6/Red Edge (b6): Koefisien -0.924

3. Temuan Penting
Data Tanah (Laboratorium):
Fosfor adalah prediktor dominan untuk nitrogen, yang masuk akal secara agronomi karena kedua nutrisi ini sering berkorelasi dalam sistem tanah subur dan aplikasi pupuk NPK. Nutrisi lain seperti Mg, Mn, dan Cu juga berperan signifikan, sementara K, Fe, dan B tidak menunjukkan pengaruh yang signifikan.
Remote Sensing:
Hanya 4 dari 22 fitur remote sensing yang signifikan. Band SWIR dan Red Edge (b7, b6) serta NIR (b8) memberikan informasi berguna tentang karakteristik vegetasi dan tanah. Yang mengejutkan, seluruh data radar (Sigma_VV, Sigma_VH, gamma, beta) tidak berkontribusi signifikan terhadap prediksi nitrogen.
4. Masalah yang Teridentifikasi
Multicollinearity Parah:
Condition number sebesar 2.98e+05 menunjukkan masalah multikolinearitas yang serius. Ini terjadi karena banyak fitur yang saling berkorelasi tinggi, terutama band spektral yang berdekatan dan berbagai transformasi data radar.
Fitur Redundan:
23 dari 30 fitur (76%) tidak signifikan secara statistik (p > 0.05), menunjukkan banyak informasi redundan dalam model. Ini termasuk hampir semua data radar dan sebagian besar band spektral visible.
Pelanggaran Asumsi:
Uji Omnibus dan Jarque-Bera menunjukkan residual tidak terdistribusi normal (p < 0.001), yang dapat mempengaruhi akurasi confidence intervals meskipun tidak mengurangi kemampuan prediksi model.
5. Implikasi Praktis
Untuk aplikasi operasional seperti pemetaan kesuburan tanah atau rekomendasi pemupukan, model ini sudah cukup baik dengan akurasi 76-78%. Namun, model dapat disederhanakan drastis dengan hanya menggunakan 8 fitur signifikan tanpa kehilangan banyak akurasi.
Data laboratorium (terutama fosfor) lebih penting daripada data remote sensing untuk prediksi nitrogen. Data radar dapat dihilangkan sepenuhnya karena tidak memberikan nilai tambah.
6. Rekomendasi
Jangka Pendek:

Buat model reduced dengan hanya 8 fitur signifikan
Eliminasi semua fitur dengan p-value > 0.10
Lakukan VIF analysis untuk mengatasi multicollinearity

Jangka Menengah:

Implementasi Ridge atau Lasso regression untuk regularisasi
Pertimbangkan feature engineering seperti rasio N/P atau indeks vegetasi custom
Standardisasi fitur menggunakan StandardScaler

Jangka Panjang:

Eksplorasi model ensemble seperti Random Forest atau XGBoost yang lebih robust terhadap multicollinearity dan dapat menangkap hubungan non-linear
Cross-validation untuk memastikan model tidak overfit
Validasi spatial untuk memastikan model dapat generalisasi ke lokasi baru

7. Penilaian Akhir
Model menunjukkan performa yang baik secara keseluruhan tetapi memerlukan perbaikan dalam hal efisiensi fitur dan penanganan multicollinearity. Dengan simplifikasi dan optimasi, model dapat mencapai performa yang sama atau lebih baik dengan kompleksitas yang jauh lebih rendah. Grade: B+ (baik tetapi perlu refinement).